### PR TITLE
Changing the display of fatal warnings

### PR DESCRIPTION
**Pull request by v-gb**
This pull request changes the display of warnings that are errors, for instance from:

```
File "a.ml", line 1, characters 6-7:
Warning 27: unused variable x.
File "tmp.ml", line 1:
Error: Some fatal warnings were triggered (1 occurrences)
```

to

```
File "a.ml", line 1, characters 6-7:
Error: Warning 27: unused variable x.
```

The main motivation is that the message about fatal warnings has a location that does not respect the line directives in the source file, which is my case means random filenames show up in errors.

This change also means one can tell from the output what warnings are errors, which could be useful either for people or for parsing the output.

Finally, it makes for a shorter error message in the common case where the code is not full of warnings.

### CODE REVIEW

File: parsing/location.ml [hunk 1]
```diff
@@ -423,10 +438,8 @@ let () =
       | Sys_error msg ->
           Some (errorf ~loc:(in_file !input_name)
                 "I/O error: %s" msg)
-      | Warnings.Errors n ->
-          Some
-            (errorf ~loc:(in_file !input_name)
-             "Some fatal warnings were triggered (%d occurrences)" n)
+
+      | Warnings.Errors _ -> Some non_displayed_error
```
Comment (alainfrisch) [inline parsing/location.ml]: This seems a bit of a hack. What about capturing Warning.Errors directly in Location.report_exception (and just do nothing)?

File: testsuite/tools/expect_test.ml [hunk 2]
```diff
@@ -137,23 +137,10 @@ module Compiler_messages = struct
...
-      Format.fprintf ppf "Warning %a@." Warnings.print w
-    end
-
   let capture ppf ~f =
     Misc.protect_refs
-      [ R (Location.formatter_for_warnings , ppf            )
-      ; R (Location.warning_printer        , warning_printer)
-      ; R (Location.error_reporter         , error_reporter )
+      [ R (Location.formatter_for_warnings , ppf)
+      ; R (Location.printer                , print_loc)
```
Comment (alainfrisch) [inline testsuite/tools/expect_test.ml:144]: Do we want to keep those old hooks?

Comment (v-gb) [reply testsuite/tools/expect_test.ml:144]: The hooks were added in https://github.com/ocaml/ocaml/pull/171, so I don't think this refactoring means we can remove them.

File: parsing/location.ml (hunk 1 above)
Comment (v-gb) [reply parsing/location.ml]: I started this way but there are two things that were a bit awkward:
- it doesn't allow to define more exceptions that don't need to be displayed
- Location.error_of_exn is exposed

ocamldoc can receive Warnings.Error but it uses Location.error_of_exn for instance, so it'd report some weird errors. But in fact, it doesn't even report warnings in mli because it's missing a call to Typecore.force_delayed_checks, and it doesn't stop on fatal warnings in ml because it's missing a call to Warnings.check_fatal (why is it even reporting warnings anyway?).

Looking at the uses of Location.error_of_exn, I think they are all worse than using Location.report_exception, so I guess I'll change that, and then it'd make sense to do what you say.

Comment (v-gb) [reply parsing/location.ml]: I think Ast_mapper.ext_of_exn would behave weirdly if I did as you say. It would rethrow the exception, and the exception would be caught at the further down, instead of being converted into an error and sent to the compiler as the other exceptions.

**Comment by mshinwell** [comment]
@alainfrisch I think we're waiting on your approval as to @sliquister 's latest comments.

File: parsing/location.mli [hunk 3]
```diff
@@ -104,9 +107,6 @@ type error =
 
 exception Error of error
 
-val print_error_prefix: formatter -> unit -> unit
```
Comment (alainfrisch) [inline parsing/location.mli]: I'm not against removing this function from the API, but is this related to the current proposal?

File: parsing/location.ml (hunk 1 above)
Comment (alainfrisch) [reply parsing/location.ml]: Was about the following: drop the Warnings.Error exception constructor altogether, have Warnings.check_fatal returns a boolean, and explicitly adapt all driver code to behave as they do currently (either exiting the process, or continuing in the case of the toplevel) when fatal warnings have been reported?  It's a much more invasive change, but it will also paves the way to a nicer treatment of reporting multiple errors in general (e.g. adapting the type-checker to report errors without failing necessarily at the first one).

**Comment by alainfrisch** [comment]
>`Error: Warning 27:`

The multiple colons do not look super nice.  What about `Fatal warning 27: ...`, or `Error (warning 27): ...`.

File: parsing/location.mli (hunk 3 above)
Comment (v-gb) [reply parsing/location.mli]: I removed the only use of this (in the expect test tool), so it seems appropriate to stop exposing it too.

**Comment by v-gb** [comment]
`Fatal warning 27: ..` was the first thing I tried, but I don't really like it: the "fatal" in fatal warning and fatal error don't mean the same thing so it's awkward. If it's an error, I'd rather print "Error".  `Error (warning 27):` sounds fine though, so I used that.

File: parsing/location.ml (hunk 1 above)
Comment (v-gb) [reply parsing/location.ml]: I'd rather have the compiler behave more like a library and less like a standalone program, so I don't really want to add calls to exit.
I don't see how the refactoring you suggest would help with reporting multiple type errors: it sounds like reporting such errors the same way as warning would work fine.

If the comparison with the empty string is what you want to remove, maybe renaming `Warnings.Errors` to `Location.Already_displayed` and changing `Location.error_of_exn` to have type `exn -> [ Ok of error | Already_displayed ] option` would work. This way `Ast_mapper.ext_of_exn` can serialize the Already_displayed case, and rethrow on the other side, and `Location.report_exception` would behave as you wrote earlier.

Comment (v-gb) [reply parsing/location.ml]: Any opinion on that last change, where I did what I described above?

**Comment by mshinwell** [comment]
@alainfrisch Are you happy with this patch now that the colon issue has been fixed?
@sliquister Please rebase and add a Changes entry.

**Comment by v-gb** [comment]
@mshinwell I made the changes.

I would squash the four middle commits too, but I'll wait until no one is asking for further changes.

File: testsuite/tools/expect_test.ml (hunk 2 above)
Comment (alainfrisch) [reply testsuite/tools/expect_test.ml:144]: Ok.

**Comment by mshinwell** [comment]
Silence from @alainfrisch taken as assent, given the nature of this patch, and previous comments.
@sliquister Please rebase again (sorry) and ping me to merge.

**Comment by v-gb** [comment]
@mshinwell I rebased and squashed.
